EIPS/eip-remove-codesize-limit.md

```diff
@@ -0,0 +1,53 @@
+---
+title: Remove Contract Code Size Limit
+description: Removes the contract code size limit introduced in EIP-170 and adds a gas metering to code loading
+author: Charles Cooper (@charles-cooper)
+discussions-to: https://ethereum-magicians.org/t/eip-remove-contract-size-limit/23156
+status: Draft
+type: Standards Track
+category: Core
+created: 2025-03-14
+requires: 170
+---
+
+## Abstract
+
+This EIP removes the hard contract code size limit of 24KB (24576 bytes) introduced in [EIP-170](./eip-170.md) and replaces it with a gas-based mechanism. It introduces a gas cost of 2 gas per (32 byte) word for contract code exceeding 24KB, allowing deployment of contracts of any size while preventing DoS attacks through appropriate gas metering.
+
+## Motivation
+
+EIP-170 introduced a 24KB contract code size limit to prevent potential DoS attacks, as large contract code requires O(n) resource cost in terms of disk reads, VM preprocessing, and Merkle proof sizes, all of which are not directly compensated by gas fees. However, this hard limit restricts legitimate use cases for large contracts.
+
+Rather than imposing an arbitrary size limit, this EIP proposes a gas-based solution that allows contracts of any size while ensuring that users loading very large contracts pay gas proportional to the additional resources they consume. This approach aligns with Ethereum's gas model philosophy of paying for the resources consumed.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+1. Remove the contract code size limit of 24KB (`0x6000` bytes) introduced in EIP-170.
+2. Change the gas schedule for opcodes which load code. Specifically, the `CALL`, `STATICCALL`, `DELEGATECALL`, `CALLCODE`, and `EXTCODESIZE` opcodes are modified so that `ceil32(excess_contract_size) * 2 // 32` gas is deducted, where `excess_contract_size = max(0, contract_size - 0x6000)`. (Cf. initcode metering: [EELS](https://github.com/ethereum/execution-specs/blob/1a587803e3e698407d204888b02342393f8b4fe5/src/ethereum/cancun/vm/gas.py#L269))
```

Comment:
Do I understand correctly that with this change the gas cost of e.g. `EXTCODESIZE` is unbounded?

Reply:
yes. this is to reflect the cost of loading the full contract code from disk and adding to the witness. in practice, this is bounded by the max contract size you could deploy with current gas limits, which is about 35 million gas / 200 gas per byte, or 170KB.

also as i mentioned, the cost for EXTCODESIZE and EXTCODECOPY could be tweaked. iiuc the bytecode is stored on disk with RLP encoding. you can peek at the first few bytes of the RLP-encoded bytecode to find out the size. but i'm not sure this optimization is worth it.

Reply:
oh, and i should update the wording to clarify that this only applies to the access cost for cold contracts. it does not apply to contracts which have already been accessed in the transaction.